pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java

```diff
@@ -90,8 +90,25 @@ public BlobInputStream(LargeObject lo, int bsize, long limit) {
     // The very first read multiplies the last buffer size by two, so we divide by two to get
     // the first read to be exactly the initial buffer size
     this.lastBufferSize = INITIAL_BUFFER_SIZE / 2;
+
+    long maxBlobSize;
+    try {
+      this.absolutePosition = lo.tell64();
+      maxBlobSize = 4L * 1024 * 1024 * 1024 * 1024 - 2 * 1024;
```

Comment:
Can you provide a link or explanation for this math ? I realize the docs say 4TB. Link to that would be good

Reply:
The docs indeed say 4TB, but you cannot actually get that big (with PG14 anyway).
I don't know why the real limit is 2kB less than 4TB.

Reply:
What is the point of having the math in the driver? What if PG 19 lifts the limit to 4 exabytes?

Reply:
How do you suggest determining the correct value otherwise?

If the limit is raised to 4 exabytes, then the `LargeObject` API will also have to change to take a value larger than a `long`, so all this code would need changing to support it anyway.

Reply:
Signed Long.MAX_VALUE is 9223372036854775807L (9 exabytes), and unsigned long is ~18 exabytes, so the current `seek64(long pos` is more-or-less enough for multi-exabyte long values. I'm not sure what is the point of artificially limiting the values at the driver level

Reply:
> Signed Long.MAX_VALUE is 9223372036854775807L (9 exabytes)

That's 8 EiB, but you're right, I miscounted the digits.

> what is the point of artificially limiting the values at the driver level

If you don't then it throws. If you prefer that behaviour then I can remove `maxBlobSize`, but that would be a breaking change for `skip()` unless we also catch and check the error message.

Reply:
> Can you provide a link or explanation for this math ? I realize the docs say 4TB. Link to that would be good

Here is a link that you might be able to use:

https://www.enterprisedb.com/postgres-tutorials/postgresql-toast-and-working-blobsclobs-explained

It seems crazy, but I think the only mention of the 4TB upper limit on blobs on postgresql.org is in the release notes for 9.3 under "E.26.3.5. Data Types": https://www.postgresql.org/docs/9.3/release-9-3.html

Reply:
@esabol: https://www.postgresql.org/docs/16/lo-intro.html

The problem is that it's _not_ 4 TB (TiB). It's slightly less.

Reply:
> The problem is that it's _not_ 4 TB (TiB). It's slightly less.

I'm aware of that. @davecramer asked for a link (to be added as a comment in the code), so I thought I'd offer what I could find.

Reply:
I found it: https://github.com/postgres/postgres/blob/REL_16_2/src/include/storage/large_object.h#L76

That's `(2^31 - 1) * 2048 == 4 * 1024^4 - 2048`

I'll change the code here to match, with a link.